typing/typecore.ml

```diff
@@ -1218,28 +1218,29 @@ let disambiguate_lid_a_list loc closed env usage expected_type lid_a_list =
       (Warnings.Name_out_of_scope (!w_scope_ty, List.rev !w_scope, true));
   lbl_a_list
 
-let rec find_record_qual = function
-  | [] -> None
-  | ({ txt = Longident.Ldot (modname, _) }, _) :: _ -> Some modname
-  | _ :: rest -> find_record_qual rest
+let find_record_qual lid_a_list =
+  let rec loop accu = function
+    | [] -> lid_a_list
+    | ({txt = Longident.Ldot (modname, _); _}, _ as lid_a) :: rest ->
+        lid_a ::
+        List.map (fun (lid, a as lid_a) ->
+            match lid.txt with
+            | Longident.Lident s ->
+                {lid with txt = Longident.Ldot (modname, s)}, a
+            | _ -> lid_a
+          ) (List.rev_append accu rest)
```

Comment:
It would be nice to have a comment that explains why we are doing this:
- if at least one field is explicitly qualified, we qualify all fields to reduce field-name ambiguity
- we reorder the explicitly-qualified field before the other, to ensure that potential error messages related to the qualification will be raised on this field first rather than the "ghost" longidents we produced
- the reordering does not change the program semantics because our caller `type_label_a_list` sorts the field by type-declaration order anyway